File: trunk/src/core/src/main/java/org/semachina/jena/datatype/factory/DateTimeFactory.java
package org.semachina.jena.datatype.factory;

import com.hp.hpl.jena.vocabulary.XSD;
import org.joda.time.DateTime;
import org.joda.time.format.ISODateTimeFormat;
import org.semachina.jena.impl.SimpleRDFDatatypeImpl;

import java.util.Date;

public class DateTimeFactory extends SimpleRDFDatatypeImpl<DateTime> {

    public DateTimeFactory() {
        super(XSD.dateTime.getURI(), DateTime.class);
    }

    @Override
    public DateTime parseLexicalForm(String lexicalForm) throws Exception {
        return ISODateTimeFormat.dateTime().parseDateTime(lexicalForm);
    }

    @Override
    public String toLexicalForm(DateTime cast) throws Exception {
        return ISODateTimeFormat.dateTime().print(cast);
    }

    @Override
    public Object cannonicalise(Object value) {
        if (value instanceof Date) {
            return new DateTime(((Date) value).getTime());
        }
        return super.cannonicalise(value);
    }
}

File: src/core/src/main/scala/org/semachina/jena/datatype/factory/DateTimeFactory.scala
package org.semachina.jena.datatype.factory

import com.hp.hpl.jena.vocabulary.XSD
import org.joda.time.DateTime
import org.joda.time.format.ISODateTimeFormat
import org.semachina.jena.datatype.SimpleRDFDatatype
import java.util.Date

class DateTimeFactory extends SimpleRDFDatatype[DateTime] (
  XSD.dateTime.getURI,
  { lexicalForm: String => ISODateTimeFormat.dateTime.parseDateTime(lexicalForm) },
  { cast: DateTime => ISODateTimeFormat.dateTime.print(cast) } ) {

  override def cannonicalise(value: AnyRef): DateTime = {
    if (value.isInstanceOf[Date]) {
      return new DateTime((value.asInstanceOf[Date]).getTime)
    }
    return super.cannonicalise(value)
  }
}